From 842ecae7678e62d97e7a9ca6efbe821444720e41 Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Mon, 14 Dec 2020 11:41:59 -0800
Subject: [PATCH 15/15] hostapd: Netlink changes for RNR offsets

This patch adds new nested netlink attribute NL80211_ATTR_EMA_RNR_ELEMS
used to send reduced neighbor report (RNR) elements to kernel when EMA
is enabled.
This attribute includes the count of RNR elements and data at each index.
While generating EMA beacons, kernel will include RNR group at a given
index along with MBSSID group.
Last element, if present, has RNR data common for all EMA beacons.

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 src/drivers/driver_nl80211.c | 19 +++++++++++++++++++
 src/drivers/nl80211_copy.h   |  2 ++
 2 files changed, 21 insertions(+)

--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -4577,6 +4577,25 @@ static int nl80211_mbssid(struct nl_msg
 			offs[i]);
 		nla_nest_end(msg, elems);
 	}
+
+	if (!params->ema)
+		return 0;
+
+	if (params->rnr_elem_count && params->rnr_elem_len &&
+	    params->rnr_elem_offset && *params->rnr_elem_offset) {
+		u8 i, **offs = params->rnr_elem_offset;
+
+		elems = nla_nest_start(msg, NL80211_ATTR_EMA_RNR_ELEMS);
+		if (!elems)
+			goto fail;
+
+		for (i = 0; i < params->rnr_elem_count - 1; i++)
+			nla_put(msg, i + 1, offs[i + 1] - offs[i], offs[i]);
+		nla_put(msg, i + 1,
+			*offs + params->rnr_elem_len- offs[i],
+			offs[i]);
+		nla_nest_end(msg, elems);
+	}
 	return 0;
 
 fail:
--- a/src/drivers/nl80211_copy.h
+++ b/src/drivers/nl80211_copy.h
@@ -3109,6 +3109,8 @@ enum nl80211_attrs {
 
 	NL80211_ATTR_6G_REG_POWER_MODE,
 
+	NL80211_ATTR_EMA_RNR_ELEMS,
+
 	/* add attributes here, update the policy in nl80211.c */
 
 	__NL80211_ATTR_AFTER_LAST,
